### PR TITLE
EID-251: Update Hub Transformers Factory

### PR DESCRIPTION
**Pull request by adityapahuja**
to be able to transform from eIDAS Attribute Query Request to SAML Attribute Query.

Authors: @adityapahuja

### CODE REVIEW

File: hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        DateTime issueInstant,
+        PersistentId persistentId, // FIXME - change to use new PersistenceId with equal and hashcode at both places
+        URI assertionConsumerServiceUrl,
+        String authnRequestIssuerEntityId,
+        String encryptedIdentityAssertion,
+        AuthnContext authnContext,
+        Optional<HubAssertion> cycle3AttributeAssertion,
+        Optional<List<UserAccountCreationAttribute>> userAccountCreationAttributes) {
+
+        super(requestId, issuer, issueInstant, null, persistentId, assertionConsumerServiceUrl, authnRequestIssuerEntityId);
```
Comment (TimWSpence) [inline hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java]: Why is this null?

File: hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package uk.gov.ida.saml.hub.domain;
```
Comment (TimWSpence) [inline hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java]: Why do we have a copy of this in saml-test-utils (as well as the one in hub saml)?

File: hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java (hunk 1 above)
Comment (adityapahuja) [reply hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java]: It is a destination field and this field is not used. No idea why destination field is added to BaseHubAttributeQueryRequest class. HubAttributeQueryRequest that extends BaseHubAttributeQueryRequest has null too.

File: hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            .withPersistentId(persistentId)
+            .build();
+
+        AttributeQuery transformedQuery = transformer.apply(hubEidasAttributeQueryRequest);
+
+        assertThat(transformedQuery.getID()).isEqualTo(hubEidasAttributeQueryRequest.getId());
+        assertThat(transformedQuery.getSubject().getNameID().getValue()).isEqualTo(persistentId.getNameId());
+        assertThat(transformedQuery.getIssuer().getValue()).isEqualTo(hubEidasAttributeQueryRequest.getIssuer());
+        assertThat(transformedQuery.getVersion()).isEqualTo(SAMLVersion.VERSION_20);
+        new XmlObjectToBase64EncodedStringTransformer<>().apply(transformedQuery);
```
Comment (TimWSpence) [inline hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java]: What is this line doing? My only guess is that it asserts that it is valid xml but if so, I would prefer for this to be explicit

File: hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        AttributeQuery transformedQuery = transformer.apply(hubEidasAttributeQueryRequest);
+
+        List<XMLObject> unknownXMLObjects = transformedQuery.getSubject().getSubjectConfirmations().get(0).getSubjectConfirmationData().getUnknownXMLObjects(Assertion.DEFAULT_ELEMENT_NAME);
+        assertThat(unknownXMLObjects.size()).isEqualTo(1);
+        assertThat(((Assertion) unknownXMLObjects.get(0)).getID()).isEqualTo(cycle3DataAssertion.getId());
+    }
+
+    @Test
+    public void shouldIncludeEncryptedIdentityAssertionOnly() throws Exception {
+        final String encryptedIdentityAssertion = ENCRYPTED_IDENTITY_ASSERTION;
```
Comment (TimWSpence) [inline hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java]: Why make a final copy of a constant?

File: hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java (hunk 2 above)
Comment (adityapahuja) [reply hub-saml-test-utils/src/main/java/uk/gov/ida/saml/hub/domain/HubEidasAttributeQueryRequest.java]: I cannot find this class in saml-test-utils. Classes in saml-test-utils cannot be used for production and are used for testing only.

File: hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java (hunk 4 above)
Comment (adityapahuja) [reply hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java]: Good spot. I used refactor to extract this variable into constant field. I have removed it.

File: hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java (hunk 3 above)
Comment (adityapahuja) [reply hub-saml/src/test/java/uk/gov/ida/saml/hub/transformers/outbound/HubEidasAttributeQueryRequestToSamlAttributeQueryTransformerTest.java]: I cloned `HubAttributeQueryRequestToSamlAttributeQueryTransformerTest.java` to create `HubAttributeQueryRequestToSamlAttributeQueryTransformerTest.java`. This line should not be there. I have removed it.

**Pull request merged**